streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -384,18 +384,20 @@ public void foreach(final ForeachAction<? super K, ? super V> action) {
 
     @Override
     public KStream<K, V> through(final String topic) {
-        return through(topic, Produced.with(null, null, null));
+        return through(topic, Produced.with(keySerde, valSerde, null));
     }
 
     @Override
     public KStream<K, V> through(final String topic, final Produced<K, V> produced) {
         final ProducedInternal<K, V> producedInternal = new ProducedInternal<>(produced);
+        producedInternal.withKeySerde(producedInternal.keySerde() != null ? producedInternal.keySerde() : keySerde);
```

Comment:
This is good.

nit: maybe clearer as `if (producedInternal.keySerde() == null) { producedInternal.withKeySerde(keySerde); }`, my gut feeling on `producedInternal.withKeySerde(producedInternal.keySerde())` calls are a bit awkward. Ditto below.

Reply:
Good call.